File: services/plesk-mail-api/pleskClient.mjs
/**
 * Plesk Obsidian XML API Client — Node.js (ESM)
 *
 * Production-ready with retry/backoff, structured logging, TLS enforcement.
 * Authentication: HTTP_AUTH_LOGIN / HTTP_AUTH_PASSWD headers.
 *
 * Dependency: fast-xml-parser ^5
 */

import https from "node:https";
import fs from "node:fs";
import { XMLParser } from "fast-xml-parser";

const PACKET_VERSION = "1.6.9.1";
const parser = new XMLParser({ ignoreAttributes: false, trimValues: true });

// ── Helpers ────────────────────────────────────────────────────

function escapeXml(str) {
  return str
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

function redactPassword(xml) {
  return xml.replace(
    /(<password>\s*<value>)([\s\S]*?)(<\/value>)/g,
    "$1***REDACTED***$3"
  );
}

function log(level, msg, ...args) {
  const ts = new Date().toISOString();
  console[level](`${ts} [plesk] ${msg}`, ...args);
}

// ── Error ──────────────────────────────────────────────────────

export class PleskAPIError extends Error {
  constructor(errcode, errtext) {
    super(`Plesk API error ${errcode}: ${errtext}`);
    this.name = "PleskAPIError";
    this.errcode = errcode;
    this.errtext = errtext;
  }
}

// ── Client ─────────────────────────────────────────────────────

export class PleskClient {
  #url;
  #headers;
  #timeout;
  #maxRetries;
  #backoffMs;
  #agentOptions;

  /**
   * @param {object} opts
   * @param {string} opts.host - Plesk hostname or IP
   * @param {string} opts.login - HTTP_AUTH_LOGIN value
   * @param {string} opts.password - HTTP_AUTH_PASSWD value
   * @param {number} [opts.port=8443]
   * @param {number} [opts.timeout=30000] - ms
   * @param {number} [opts.maxRetries=3]
   * @param {number} [opts.backoffMs=1000]
   * @param {boolean} [opts.rejectUnauthorized=true]
   * @param {string} [opts.caCertPath] - Path to CA cert for self-signed
   */
  constructor({
    host,
    login,
    password,
    port = 8443,
    timeout = 30000,
    maxRetries = 3,
    backoffMs = 1000,
    rejectUnauthorized = true,
    caCertPath,
  }) {
    this.#url = new URL(
      `https://${host}:${port}/enterprise/control/agent.php`
    );
    this.#headers = {
      HTTP_AUTH_LOGIN: login,
      HTTP_AUTH_PASSWD: password,
      "Content-Type": "text/xml",
    };
    this.#timeout = timeout;
    this.#maxRetries = maxRetries;
    this.#backoffMs = backoffMs;
    this.#agentOptions = { rejectUnauthorized };

    if (caCertPath) {
      this.#agentOptions.ca = fs.readFileSync(caCertPath);
    }

    log("info", `Initialized for ${host}:${port}`);
  }

  // ── Internal ───────────────────────────────────────────────

  #buildPacket(innerXml) {
    return `<?xml version="1.0" encoding="UTF-8"?><packet version="${PACKET_VERSION}">${innerXml}</packet>`;
  }

  async #send(innerXml) {
    const body = this.#buildPacket(innerXml);
    log("debug", "Request: %s", redactPassword(body).slice(0, 500));

    let lastErr;
    for (let attempt = 0; attempt <= this.#maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = this.#backoffMs * 2 ** (attempt - 1);
        log("warn", `Retry ${attempt}/${this.#maxRetries} in ${delay}ms`);
        await new Promise((r) => setTimeout(r, delay));
      }

      try {
        const t0 = performance.now();
        const text = await this.#request(body);
        const elapsed = ((performance.now() - t0) / 1000).toFixed(2);
        log("debug", `Response in ${elapsed}s (${text.length} bytes)`);

        const parsed = parser.parse(text);
        if (!parsed.packet) {
          throw new Error("Missing <packet> root in response");
        }
        return parsed.packet;
      } catch (err) {
        lastErr = err;
        // Don't retry Plesk API errors or client errors
        if (err instanceof PleskAPIError) throw err;
        if (err.statusCode && err.statusCode < 500) throw err;
        log("error", `Attempt ${attempt} failed: ${err.message}`);
      }
    }
    throw lastErr;
  }

  #request(body) {
    return new Promise((resolve, reject) => {
      const options = {
        method: "POST",
        hostname: this.#url.hostname,
        port: this.#url.port,
        path: this.#url.pathname,
        headers: {
          ...this.#headers,
          "Content-Length": Buffer.byteLength(body),
        },
        timeout: this.#timeout,
        ...this.#agentOptions,
      };

      const req = https.request(options, (res) => {
        const chunks = [];
        res.on("data", (chunk) => chunks.push(chunk));
        res.on("end", () => {
          const data = Buffer.concat(chunks).toString("utf-8");
          if (res.statusCode >= 400) {
            const err = new Error(`HTTP ${res.statusCode}`);
            err.statusCode = res.statusCode;
            return reject(err);
          }
          resolve(data);
        });
      });

      req.on("error", reject);
      req.on("timeout", () => {
        req.destroy();
        reject(new Error("Request timed out"));
      });
      req.write(body);
      req.end();
    });
  }

  #checkResult(packet, path) {
    const result = this.#dig(packet, path) ?? this.#dig(packet, "result");
    if (!result) {
      throw new PleskAPIError("0", "No result in response");
    }
    if (result.status !== "ok") {
      throw new PleskAPIError(
        String(result.errcode ?? "unknown"),
        result.errtext ?? "Unknown error"
      );
    }
    return result;
  }

  #dig(obj, dotPath) {
    return dotPath.split(".").reduce((o, k) => o?.[k], obj);
  }

  // ── Site ID ────────────────────────────────────────────────

  async getSiteId(domain) {
    const xml = `<site><get><filter><name>${escapeXml(domain)}</name></filter><dataset><gen_info/></dataset></get></site>`;
    const packet = await this.#send(xml);
    const result = this.#checkResult(packet, "site.get.result");
    const id = result.id ?? result.data?.gen_info?.id;
    if (!id) throw new PleskAPIError("0", `No site-id for domain ${domain}`);
    log("info", `Resolved ${domain} → site-id ${id}`);
    return Number(id);
  }

  // ── Mail CRUD ──────────────────────────────────────────────

  async createMailbox(siteId, name, password) {
    const xml =
      `<mail><create><filter>` +
      `<site-id>${siteId}</site-id>` +
      `<mailname><name>${escapeXml(name)}</name>` +
      `<mailbox><enabled>true</enabled></mailbox>` +
      `<password><value>${escapeXml(password)}</value><type>plain</type></password>` +
      `</mailname></filter></create></mail>`;
    const packet = await this.#send(xml);
    const result = this.#checkResult(packet, "mail.create.result");
    log("info", `Created mailbox '${name}' (id=${result.id}) on site ${siteId}`);
    return Number(result.id);
  }

  async updateMailboxPassword(siteId, name, newPassword) {
    const xml =
      `<mail><update><set><filter>` +
      `<site-id>${siteId}</site-id>` +
      `<mailname><name>${escapeXml(name)}</name>` +
      `<password><value>${escapeXml(newPassword)}</value><type>plain</type></password>` +
      `</mailname></filter></set></update></mail>`;
    const packet = await this.#send(xml);
    this.#checkResult(packet, "mail.update.result");
    log("info", `Updated password for '${name}' on site ${siteId}`);
    return true;
  }

  async enableMailbox(siteId, name, enabled = true) {
    const flag = enabled ? "true" : "false";
    const xml =
      `<mail><update><set><filter>` +
      `<site-id>${siteId}</site-id>` +
      `<mailname><name>${escapeXml(name)}</name>` +
      `<mailbox><enabled>${flag}</enabled></mailbox>` +
      `</mailname></filter></set></update></mail>`;
    const packet = await this.#send(xml);
    this.#checkResult(packet, "mail.update.result");
    log("info", `Set '${name}' enabled=${flag} on site ${siteId}`);
    return true;
  }

  async deleteMailbox(siteId, name, { idempotent = false } = {}) {
    const xml =
      `<mail><remove><filter>` +
      `<site-id>${siteId}</site-id>` +
      `<mailname><name>${escapeXml(name)}</name></mailname>` +
      `</filter></remove></mail>`;
    try {
      const packet = await this.#send(xml);
      this.#checkResult(packet, "mail.remove.result");
    } catch (err) {
      if (idempotent && err instanceof PleskAPIError && err.errcode === "1013") {
        log("info", `Mailbox '${name}' already deleted (idempotent)`);
        return true;
      }
      throw err;
    }
    log("info", `Deleted mailbox '${name}' on site ${siteId}`);
    return true;
  }

  async listMailboxes(siteId) {
    const xml =
      `<mail><get_info><filter>` +
      `<site-id>${siteId}</site-id>` +
      `</filter><mailbox/></get_info></mail>`;
    const packet = await this.#send(xml);
    const results = packet?.mail?.get_info?.result;
    if (!results) return [];
    const arr = Array.isArray(results) ? results : [results];
    const mailboxes = arr
      .filter((r) => r.status === "ok" && r.mailname?.name)
      .map((r) => ({
        name: r.mailname.name,
        enabled: String(r.mailname?.mailbox?.enabled) === "true",
        siteId,
      }));
    log("info", `Listed ${mailboxes.length} mailboxes on site ${siteId}`);
    return mailboxes;
  }
}

File: services/plesk-mail-api/plesk_client.py
"""Plesk Obsidian XML API Client — Production-ready.

Supports mail provisioning operations (create, update, delete, list)
with retry logic, structured logging, and TLS enforcement.

Authentication: HTTP_AUTH_LOGIN / HTTP_AUTH_PASSWD headers only.
"""

from __future__ import annotations

import html
import logging
import re
import time
import defusedxml.ElementTree as ET
from dataclasses import dataclass
from typing import Self

import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

logger = logging.getLogger("plesk_client")

# ── Redaction helpers ───────────────────────────────────────────


def redact_headers(headers: dict[str, str]) -> dict[str, str]:
    """Return a copy with HTTP_AUTH_PASSWD masked."""
    out = dict(headers)
    if "HTTP_AUTH_PASSWD" in out:
        out["HTTP_AUTH_PASSWD"] = "***REDACTED***"
    return out


def redact_xml_passwords(xml: str) -> str:
    """Mask <value>…</value> content inside <password> blocks."""
    return re.sub(
        r"(<password>\s*<value>)(.*?)(</value>)",
        r"\1***REDACTED***\3",
        xml,
        flags=re.DOTALL,
    )


# ── Data classes ────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class PleskMailbox:
    name: str
    enabled: bool
    site_id: int


# ── Exceptions ──────────────────────────────────────────────────


class PleskAPIError(Exception):
    """Raised when the Plesk XML API returns status=error."""

    def __init__(self, errcode: str, errtext: str) -> None:
        self.errcode = errcode
        self.errtext = errtext
        super().__init__(f"Plesk API error {errcode}: {errtext}")


# ── Client ──────────────────────────────────────────────────────


class PleskClient:
    """Low-level Plesk Obsidian XML API client.

    Usage::

        with PleskClient("plesk.example.com", "admin", "secret") as c:
            sid = c.get_site_id("example.com")
            c.create_mailbox(sid, "user", "P@ssw0rd!")
            for mb in c.list_mailboxes(sid):
                print(mb.name, mb.enabled)
    """

    PACKET_VERSION = "1.6.9.1"

    def __init__(
        self,
        host: str,
        login: str,
        password: str,
        *,
        port: int = 8443,
        verify_ssl: bool | str = True,
        timeout: int = 30,
        max_retries: int = 3,
        backoff_factor: float = 1.0,
    ) -> None:
        self._url = f"https://{host}:{port}/enterprise/control/agent.php"
        self._headers = {
            "HTTP_AUTH_LOGIN": login,
            "HTTP_AUTH_PASSWD": password,
            "Content-Type": "text/xml",
        }
        self._verify = verify_ssl
        self._timeout = timeout

        self._session = requests.Session()
        retry = Retry(
            total=max_retries,
            backoff_factor=backoff_factor,
            status_forcelist=[502, 503, 504],
            allowed_methods=["POST"],
        )
        self._session.mount("https://", HTTPAdapter(max_retries=retry))

        logger.info(
            "PleskClient initialized for %s:%d (verify=%s)",
            host,
            port,
            verify_ssl,
        )

    # ── Internal helpers ────────────────────────────────────────

    @staticmethod
    def _esc(value: str) -> str:
        """Escape user input for safe XML embedding."""
        return html.escape(value, quote=True)

    def _build_packet(self, inner_xml: str) -> str:
        return (
            f'<?xml version="1.0" encoding="UTF-8"?>'
            f'<packet version="{self.PACKET_VERSION}">'
            f"{inner_xml}"
            f"</packet>"
        )

    def _send(self, inner_xml: str) -> ET.Element:
        payload = self._build_packet(inner_xml)
        logger.debug("Request: %s", redact_xml_passwords(payload[:500]))

        t0 = time.monotonic()
        resp = self._session.post(
            self._url,
            headers=self._headers,
            data=payload.encode("utf-8"),
            verify=self._verify,
            timeout=self._timeout,
        )
        elapsed = time.monotonic() - t0
        logger.debug("Response %d in %.2fs (%d bytes)", resp.status_code, elapsed, len(resp.content))

        resp.raise_for_status()

        root = ET.fromstring(resp.text)
        if root.tag != "packet":
            raise ValueError(f"Unexpected root element: {root.tag}")
        return root

    def _check_result(self, root: ET.Element, path: str = ".//result") -> ET.Element:
        result = root.find(path)
        if result is None:
            raise PleskAPIError("0", "No <result> element in response")
        status_el = result.find("status")
        if status_el is None:
            raise PleskAPIError("0", "No <status> element in result")
        if status_el.text != "ok":
            errcode = result.findtext("errcode", "unknown")
            errtext = result.findtext("errtext", "Unknown error")
            raise PleskAPIError(errcode, errtext)
        return result

    # ── Lifecycle ───────────────────────────────────────────────

    def close(self) -> None:
        self._session.close()
        logger.info("PleskClient session closed")

    def __enter__(self) -> Self:
        return self

    def __exit__(self, *args: object) -> None:
        self.close()

    # ── Site ID lookup ──────────────────────────────────────────

    def get_site_id(self, domain: str) -> int:
        """Resolve a domain name to its Plesk site-id."""
        safe_domain = self._esc(domain)
        xml = (
            f"<site><get>"
            f"<filter><name>{safe_domain}</name></filter>"
            f"<dataset><gen_info/></dataset>"
            f"</get></site>"
        )
        root = self._send(xml)
        result = self._check_result(root)
        site_id = result.findtext(".//id")
        if not site_id:
            raise PleskAPIError("0", f"No site-id returned for domain {domain}")
        logger.info("Resolved domain %s → site-id %s", domain, site_id)
        return int(site_id)

    # ── Mail CRUD ───────────────────────────────────────────────

    def create_mailbox(self, site_id: int, name: str, password: str) -> int:
        """Create a new mailbox. Returns the mailbox id."""
        safe_name = self._esc(name)
        safe_pass = self._esc(password)
        xml = (
            f"<mail><create><filter>"
            f"<site-id>{site_id}</site-id>"
            f"<mailname>"
            f"<name>{safe_name}</name>"
            f"<mailbox><enabled>true</enabled></mailbox>"
            f"<password><value>{safe_pass}</value><type>plain</type></password>"
            f"</mailname>"
            f"</filter></create></mail>"
        )
        root = self._send(xml)
        result = self._check_result(root)
        mail_id = result.findtext("id", "0")
        logger.info("Created mailbox '%s' (id=%s) on site %d", name, mail_id, site_id)
        return int(mail_id)

    def update_mailbox_password(self, site_id: int, name: str, new_password: str) -> bool:
        """Update a mailbox password."""
        safe_name = self._esc(name)
        safe_pass = self._esc(new_password)
        xml = (
            f"<mail><update><set><filter>"
            f"<site-id>{site_id}</site-id>"
            f"<mailname>"
            f"<name>{safe_name}</name>"
            f"<password><value>{safe_pass}</value><type>plain</type></password>"
            f"</mailname>"
            f"</filter></set></update></mail>"
        )
        root = self._send(xml)
        self._check_result(root)
        logger.info("Updated password for mailbox '%s' on site %d", name, site_id)
        return True

    def enable_mailbox(self, site_id: int, name: str, enabled: bool = True) -> bool:
        """Enable or disable a mailbox."""
        safe_name = self._esc(name)
        flag = "true" if enabled else "false"
        xml = (
            f"<mail><update><set><filter>"
            f"<site-id>{site_id}</site-id>"
            f"<mailname>"
            f"<name>{safe_name}</name>"
            f"<mailbox><enabled>{flag}</enabled></mailbox>"
            f"</mailname>"
            f"</filter></set></update></mail>"
        )
        root = self._send(xml)
        self._check_result(root)
        logger.info("Set mailbox '%s' enabled=%s on site %d", name, flag, site_id)
        return True

    def delete_mailbox(self, site_id: int, name: str, *, idempotent: bool = False) -> bool:
        """Delete a mailbox. If idempotent=True, ignore 'not found' errors."""
        safe_name = self._esc(name)
        xml = (
            f"<mail><remove><filter>"
            f"<site-id>{site_id}</site-id>"
            f"<mailname><name>{safe_name}</name></mailname>"
            f"</filter></remove></mail>"
        )
        try:
            root = self._send(xml)
            self._check_result(root)
        except PleskAPIError as exc:
            if idempotent and exc.errcode == "1013":
                logger.info("Mailbox '%s' already deleted (idempotent)", name)
                return True
            raise
        logger.info("Deleted mailbox '%s' on site %d", name, site_id)
        return True

    def list_mailboxes(self, site_id: int) -> list[PleskMailbox]:
        """List all mailboxes for a site."""
        xml = (
            f"<mail><get_info><filter>"
            f"<site-id>{site_id}</site-id>"
            f"</filter><mailbox/></get_info></mail>"
        )
        root = self._send(xml)
        mailboxes: list[PleskMailbox] = []
        for result in root.findall(".//result"):
            if result.findtext("status") != "ok":
                continue
            name = result.findtext(".//name")
            enabled = result.findtext(".//mailbox/enabled", "false") == "true"
            if name:
                mailboxes.append(PleskMailbox(name=name, enabled=enabled, site_id=site_id))
        logger.info("Listed %d mailboxes on site %d", len(mailboxes), site_id)
        return mailboxes
